### PR TITLE
VACMS-11497: Set last-touched-by-a-human field on all published nodes.

### LINKED ISSUE #11497: Write script for last human review for CMS content
## Description

We need to parse out CMS migrator and bot-updated content from the CMS so that we might understand when the last human updated a piece of content. This will only count for published content, not those pieces in draft. This needs to be added to every node. Then you will have to save each node. 

All drafts will need to be floated to the top again (remember Linky) if new revisions are added.

## Implementation Recommendation

- No new revision for bulk scripted edits

## Acceptance Criteria
- [ ] Write script that populates new field with date/time stamp from the last human revision/edit 
- [ ] process every published node on site
- [ ] Determine implementation recommendation 
- [ ] exclude user 1 and CMS migrator
- [ ] No revision log needed
- [ ] create runbook 

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [x] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by ndouglas**
## Description

Closes #11497.

This needs more testing, and I'd prefer to close an unrelated issue first, because it's creating a lot of logspam:

Screen Shot 2022-11-25 at 12 56 12 PM

The script is split into the following files:

- `common.php` - library of shared code
- `delete_queue.php` - delete the queue and all of its items, mostly for cleanup between tests
- `load_queue.php` - load every published node in the DB into a queue for processing
- `print_full_history_of_node.php` - used to print a full revision history of the node, relevant fields, creation dates, etc.
- `print_node.php` - print a summary of the node with relevant fields, etc, _at the default revision_.
- `run_queue.php` - dequeue one chunk of nodes from the queue and update each one's default and latest revisions in place.
- `set_single.php` - print the full revision history of the node, update its revisions as appropriate, and then print the full revision history again for comparison.

## Testing done

Brute-force testing while developing.

## QA steps

So, generally, you wanna do the following:

### Pick a node and print its full history.

`drush scr ./scripts/content/vacms-11497/print_full_history_of_node.php 27263`

You should see some history like this:

```
$ ddev drush scr ./scripts/content/vacms-11497/print_full_history_of_node.php 27263
Acting as CMS Migrator [1317]
NID    | VID     | Publish | Default | UID   | Human | Created             | Touched (Set)       | Touched (Calc)      | Touched (Node)      | Touched (Will Set)  | Title                                    |
------ | ------- | ------- | ------- | ----- | ----- | ------------------- | ------------------- | ------------------- | ------------------- | ------------------- | ---------------------------------------- |
27263  | 469858  | Yes     | No      | 1203  | Yes   | 2021-07-08 17:44:56 | (not set)           | 2021-07-08 17:44:56 | 2021-07-08 17:44:56 | (not set)           | How do I change the name of a health ser |
27263  | 469861  | Yes     | No      | 1203  | Yes   | 2021-07-08 17:56:53 | (not set)           | 2021-07-08 17:56:53 | 2021-07-08 17:56:53 | (not set)           | How do I change the name of a health ser |
27263  | 469862  | Yes     | No      | 1203  | Yes   | 2021-07-08 18:00:20 | (not set)           | 2021-07-08 18:00:20 | 2021-07-08 18:00:20 | (not set)           | How do I change the name of a health ser |
27263  | 527768  | Yes     | No      | 197   | Yes   | 2021-09-01 16:00:21 | (not set)           | 2021-09-01 16:00:21 | 2021-09-01 16:00:21 | (not set)           | How do I change the name of a health ser |
27263  | 586334  | Yes     | No      | 1451  | Yes   | 2021-11-19 17:46:43 | (not set)           | 2021-11-19 17:46:43 | 2021-11-19 17:46:43 | (not set)           | How do I change the name of a health ser |
27263  | 626360  | Yes     | Yes     | 1451  | Yes   | 2022-03-22 10:24:25 | 2022-03-22 10:24:25 | 2022-03-22 10:24:25 | 2022-03-22 10:24:25 | 2022-03-22 10:24:25 | How do I change the name of a health ser |
27263  | 719974  | No      | No      | 3015  | Yes   | 2022-10-26 13:53:14 | 2022-10-26 13:53:14 | 2022-10-26 13:53:14 | 2022-10-26 13:53:14 | 2022-10-26 13:53:14 | How do I change the name of a health ser |
```

and a table like this:

NID    | VID     | Publish | Default | UID   | Human | Created             | Touched (Set)       | Touched (Calc)      | Touched (Node)      | Touched (Will Set)  | Title                                    |
------ | ------- | ------- | ------- | ----- | ----- | ------------------- | ------------------- | ------------------- | ------------------- | ------------------- | ---------------------------------------- |
27263  | 469858  | Yes     | No      | 1203  | Yes   | 2021-07-08 17:44:56 | (not set)           | 2021-07-08 17:44:56 | 2021-07-08 17:44:56 | (not set)           | How do I change the name of a health ser |
27263  | 469861  | Yes     | No      | 1203  | Yes   | 2021-07-08 17:56:53 | (not set)           | 2021-07-08 17:56:53 | 2021-07-08 17:56:53 | (not set)           | How do I change the name of a health ser |
27263  | 469862  | Yes     | No      | 1203  | Yes   | 2021-07-08 18:00:20 | (not set)           | 2021-07-08 18:00:20 | 2021-07-08 18:00:20 | (not set)           | How do I change the name of a health ser |
27263  | 527768  | Yes     | No      | 197   | Yes   | 2021-09-01 16:00:21 | (not set)           | 2021-09-01 16:00:21 | 2021-09-01 16:00:21 | (not set)           | How do I change the name of a health ser |
27263  | 586334  | Yes     | No      | 1451  | Yes   | 2021-11-19 17:46:43 | (not set)           | 2021-11-19 17:46:43 | 2021-11-19 17:46:43 | (not set)           | How do I change the name of a health ser |
27263  | 626360  | Yes     | Yes     | 1451  | Yes   | 2022-03-22 10:24:25 | 2022-03-22 10:24:25 | 2022-03-22 10:24:25 | 2022-03-22 10:24:25 | 2022-03-22 10:24:25 | How do I change the name of a health ser |
27263  | 719974  | No      | No      | 3015  | Yes   | 2022-10-26 13:53:14 | 2022-10-26 13:53:14 | 2022-10-26 13:53:14 | 2022-10-26 13:53:14 | 2022-10-26 13:53:14 | How do I change the name of a health ser |

or this:
Screen Shot 2022-11-25 at 1 09 08 PM

The columns are:
- **NID**: Node ID.
- **VID**: Revision ID.
- **Publish**: Whether this specific revision is published.
- **Default**: Whether this specific revision is the default revision.
- **UID**: The UID of the user who authored the revision.
- **Human**: Whether the authoring user was a human.
- **Touched (Set)**: The value of the last-touched-by-a-human field _actually set_ for the revision, or `(not set)` if it hasn't been set.
- **Touched (Calc)**: The value of the last-touched-by-a-human field _calculated_ for the revision based on the revision logic.  This will equal **Created** if the revision is updated by a human, otherwise `(not set)`.
- **Touched (Node)**: The value of the last-touched-by-a-human field as should be reported for the revision based on the per-revision logic, but using the value from the previous revision(s) if set.
- **Touched (Will Set)**: The value of the last-touched-by-a-human field that will be actually set, on the revisions that will actually be altered by the script under development.
- **Title**: (truncated)

Verify that the logic seems appropriate.
- If **Touched (Set)** is No, then the last-touched-by-a-human field is not currently set.  
- **Touched (Will Set)** indicates what that field will be set to, if we attempt a write operation.  Does that reflect the logic we intend to use?

### Update a single node and observe the changes

`drush scr ./scripts/content/vacms-11497/set_single.php 27263`

You should see a table of the node's full revision history, then messages tracing through the basic logic of whether or not to set the last-updated field.  They may look like this:

```
Updating node 27263...
Inspecting default revision 626360 for node 27263
Value already set for revision 626360 to 1647959065 (2022-03-22 10:24:25); not overwriting.
Inspecting latest revision 719974 for node 27263
Value already set for revision 719974 to 1666806794 (2022-10-26 13:53:14); not overwriting.
```

or like this:

```
Updating node 27387...
Inspecting latest & default revision 619119 for node 27387
Value already set for revision 619119 to 1645563731 (2022-02-22 16:02:11); not overwriting.
```

or like this:

```
Updating node 50654...
Inspecting latest & default revision 723581 for node 50654
Setting revision last-human date for revision 723581 to 1666892832 (2022-10-27 13:47:12).
```

or like this:

```
Updating node 39262...
Inspecting latest & default revision 724832 for node 39262
Setting revision last-human date for revision 724832 to 1664389389 (2022-09-28 14:23:09).
The default revision is not published, so we ignore it.
```

etc.

### As the script would actually be run
```bash
# Optional, if you've already done something involving the queue.
drush scr ./scripts/content/vacms-11497/delete_queue.php
drush scr ./scripts/content/vacms-11497/load_queue.php
drush scr ./scripts/content/vacms-11497/run_queue.php
```

So, the big questions are -- uh, just a sec

### The Big Questions
- this grabs all published nodes, then for each, loads the latest revision and sets its last-modified-by-a-human date.  If the default revision is distinct, the default revision is also processed similarly.
- that matches my understanding of what we covered in the meeting, but it might also be that we only want to set the last-modified-by-a-human date on the **default** revision.  I feel like we discussed both sides of this and went with both, but now I'm doubting myself and the ACs don't cover this.

### CODE REVIEW

**Review by edmund-dunn — DISMISSED** [review]
This all looks good to me. The results look consistent. The code all makes sense and is clean!

**Comment by ndouglas** [comment]
502 bad gateway to blame for the failing Cypress test 😩

File: scripts/content/vacms-11497/common.php [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+/**
+ * Load the latest revision of a node.
+ *
+ * @param int $nid
+ *   The node ID.
+ *
+ * @return \Drupal\node\NodeInterface
+ *   The latest revision of that node.
+ */
+function get_latest_node_revision(int $nid): NodeInterface {
```
Comment (swirtSJW) [inline scripts/content/vacms-11497/common.php:70]: Some of these functions are super useful outside of this specific issue and some reproduce what is already here scripts/content/script-library.php    
Can we move the ones that have reusable value to script-library.php  ?

File: scripts/content/vacms-11497/common.php [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+use Drupal\user\UserStorageInterface;
+
+define('CMS_MIGRATOR_ID', 1317);
+define('CMS_QUEUE_NAME', 'vacms_11497');
+define('CMS_CHUNK_LENGTH', 50);
+define('CMS_OUTPUT_HEADER', "%-6s | %-7s | %-7s | %-7s | %-5s | %-5s | %-19s | %-19s | %-19s | %-19s | %-19s | %-40s |\n");
+define('CMS_OUTPUT_SEPARATOR', "%'--6s | %'--7s | %'--7s | %'--7s | %'--5s | %'--5s | %'--19s | %'--19s | %'--19s | %'--19s | %'--19s | %'--40s |\n");
+define('CMS_EXCLUSION_USER_IDS', [
+  0,
+  1,
```
Comment (swirtSJW) [inline scripts/content/vacms-11497/common.php:19]: Nice

File: scripts/content/vacms-11497/common.php [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+  CMS_MIGRATOR_ID,
+]);
+
+/**
+ * Log a message.
+ *
+ * @param string $message
+ *   The message to log.
+ */
+function log_message(string $message): void {
```
Comment (swirtSJW) [inline scripts/content/vacms-11497/common.php:29]: A definite candidate for content/script-library.php  Though it might be good to name to indicate it is a terminal log message vs revisoon log or watchdog..

File: scripts/content/vacms-11497/common.php [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+  // \Drupal::logger(__FILE__)->notice($message);
+  echo $message . PHP_EOL;
+}
+
+/**
+ * Get the node storage.
+ */
+function get_node_storage(): NodeStorageInterface {
```
Comment (swirtSJW) [inline scripts/content/vacms-11497/common.php:37]: candidate scripts/content/script-library.php

File: scripts/content/vacms-11497/common.php [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+  $entity_type_manager = \Drupal::entityTypeManager();
+  return $entity_type_manager->getStorage('node');
+}
+
+/**
+ * Get the user storage.
+ */
+function get_user_storage(): UserStorageInterface {
```
Comment (swirtSJW) [inline scripts/content/vacms-11497/common.php:45]: candidate scripts/content/script-library.php

File: scripts/content/vacms-11497/common.php [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+  $entity_type_manager = \Drupal::entityTypeManager();
+  return $entity_type_manager->getStorage('user');
+}
+
+/**
+ * Switch to the CMS Migrator user.
+ */
+function switch_user(): void {
```
Comment (swirtSJW) [inline scripts/content/vacms-11497/common.php:53]: scripts/content/script-library.php   Though would be more honest to have it support a uid arg that defaults to cms migrator.

File: scripts/content/vacms-11497/common.php (hunk 1 above)
Comment (swirtSJW) [reply scripts/content/vacms-11497/common.php:70]: This one is a duplicate of script-library.php::get_node_at_latest_revision.  Your logic is better because you compartmentalized the drupal calls.  Would be good to port it to there.

File: scripts/content/vacms-11497/common.php [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+/**
+ * Load the default revision of a node.
+ *
+ * @param int $nid
+ *   The node ID.
+ *
+ * @return \Drupal\node\NodeInterface
+ *   The default revision of that node.
+ */
+function get_default_node_revision(int $nid): NodeInterface {
```
Comment (swirtSJW) [inline scripts/content/vacms-11497/common.php:85]: candidate  script-library.php

File: scripts/content/vacms-11497/common.php [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+/**
+ * Load all revisions of a node.
+ *
+ * @param int $nid
+ *   The node ID.
+ *
+ * @return \Drupal\node\NodeInterface
+ *   The default revision of that node.
+ */
+function get_all_node_revisions(int $nid): array {
```
Comment (swirtSJW) [inline scripts/content/vacms-11497/common.php:99]: candidate script-library.php

File: scripts/content/vacms-11497/common.php (hunk 1 above)
Comment (ndouglas) [reply scripts/content/vacms-11497/common.php:70]: Yep, I'll do that tomorrow 🙂

File: scripts/content/vacms-11497/common.php [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+    $revision_id = $latest->getRevisionId();
+    log_message("Inspecting latest & default revision $revision_id for node $nid");
+  }
+  update_revision($latest, $force);
+}
+
+/**
+ * Delete the queue.
+ */
+function delete_queue(): void {
```
Comment (swirtSJW) [inline scripts/content/vacms-11497/common.php:345]: I like the queue work.  I wonder if delete_queue and load_queue should go to script-library.php ? and then just 
run_queue would be what get's copied and altered in each script that needs it... or maybe run_queue accepst an argument that is the callback name like "set_node_last_human_date"

This is so powerful it would be great to have it be reusable.

**Review by swirtSJW — COMMENTED** [review]
The logging is so beautiful I think I want to cry. 

All looking great.  Just some suggestions for consolidating re-useable code.

File: scripts/content/vacms-11497/common.php (hunk 9 above)
Comment (ndouglas) [reply scripts/content/vacms-11497/common.php:260]: Yeah, I recall that the Queue API is not as ELI5 documented as it should be.  I end up having to do some reading every time I use it just because I don't remember e.g. whether a queue item is deleted if it's processed without throwing an exception or if `::deleteItem()` has to be called manually, etc.  (I feel like that might've changed from 7 -> 8, complicating the situation.)

Comment (ndouglas) [reply scripts/content/vacms-11497/common.php:270]: I didn't end up refactoring this part.  After playing with it a bit and factoring out a bit of boilerplate, all that's left is really the specific implementation details for the task at hand.  I think that's probably a sign that the Queue API's about as useful as it's gonna get, and that the real issue is just the lack of that ELI5 documentation.  I created #11733 to add some project-specific documentation on how we can use it and also to try to improve the d.o docs.

**Comment by ndouglas** [comment]
Re-ran script after latest modifications:

```
real    79m42.698s
user    57m3.361s
sys     5m18.954s
```

I'm going to go in and click around on a good number of nodes and see if I see any monkey business. 🐵

**Comment by ndouglas** [comment]
Some monkey business found, investigating.

**Comment by ndouglas** [comment]
As far as I can tell, the actual dates on revisions are being set correctly.  However, I think there's a core bug that is preventing us from minimizing the actual impact to editors as much as I'd like.

See https://www.drupal.org/project/drupal/issues/2329253 , which I think might be the bug in question.

I'm trying to edit the node in-place, as much as possible, which includes setting `isSyncing` to `TRUE`.

Screen Shot 2022-11-30 at 9 23 14 AM

However, that still seems to change the `changed` date on nodes, which is visible in the Content Admin view and others.

So we might see this:

Screen Shot 2022-11-30 at 9 24 55 AM

which makes it look as though we're running the Linky updater script again, which is a valid cause for concern.

(Admittedly, that's only when I shoot back many, many pages -- but it was alarming to me when I did so while testing, so I wanna bring it up.)

More recently, we'll see things like this:

Screen Shot 2022-11-30 at 9 26 27 AM

It seems innocuous, and it more-or-less is (which is why I didn't notice it before).  But of course, that last-modified date is not to be trusted.

Viewing the node, we see:

Screen Shot 2022-11-30 at 9 27 39 AM

which _appears_ inaccurate.  But viewing the revisions page:

Screen Shot 2022-11-30 at 9 27 50 AM

So we see that the date-last-touched-by-a-human field is set correctly, but the block in node view is not accurately representing the node revision date.

Looking at the view, we see:

Screen Shot 2022-11-30 at 9 37 23 AM

But the top date is coming in via an attachment:

Screen Shot 2022-11-30 at 9 38 11 AM

and that's actually using the Changed date, not the revision created date:

Screen Shot 2022-11-30 at 9 38 48 AM

Changing that to use the revision created date, we see a consistent sidebar block.

Screen Shot 2022-11-30 at 9 41 12 AM

So **TL;DR:** I think everything's right in the script itself but there are some confusing discrepancies that make this a little terrifying and not quite as low-disruption as we'd originally planned.

**Comment by swirtSJW** [comment]
@ndouglas I think I have the solution to the changed date being update, but I want to test it before I send you down hole of small rabbit.

**Review by swirtSJW — COMMENTED** [review]
I think the first set of code suggestion I made solves it, but I am pulling a new db to try it fresh just to make sure.

**Comment by swirtSJW** [comment]
Re-ran... that did not fix :(    Still poking at this.

**Comment by swirtSJW** [comment]
It is this bit of code in docroot/core/lib/Drupal/Core/Field/Plugin/Field/FieldType/ChangedItem.php  Lines 49-51   

```
if ($this->value == $original_value && $entity->hasTranslationChanges()) {
          $this->value = REQUEST_TIME;
        }

```

**Comment by swirtSJW** [comment]
Micorsecond tweak for the win
image


image


The changed time is NOT getting updated to the current time., which is great.

File: scripts/content/vacms-11497/common.php [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+    return;
+  }
+  // If this draft is the default and it's not published, then we're doing
+  // something weird.  We should only be running this (at this time) on nodes
+  // with at least one published revision.
+  if (!$node->isPublished() && $node->isDefaultRevision()) {
+    debug_log_message("The default revision is not published, so we ignore it.");
+    return;
+  }
+  $node->setNewRevision(FALSE);
```
[the remark this answers is not in the document]
Comment (swirtSJW) [reply scripts/content/vacms-11497/common.php:174]: Disregard.  I already pushed a related change

File: scripts/content/vacms-11497/common.php [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+  $node->setSyncing(TRUE);
```
[the remark this answers is not in the document]
Comment (swirtSJW) [reply scripts/content/vacms-11497/common.php:176]: disregard, already pushed related

**Review by swirtSJW — CHANGES_REQUESTED** [review]
I think something is off here with looking for human updates that are farther down the revision stack.

This node is not picking up the human edit /node/2273/
image

**Comment by swirtSJW** [comment]
I see the problem.  Forcing in the date to get around the new save, is also forcing in the date and revision info of the last human saved... which is not alwasy the same as the revision being saved, since we are doing latests and default.
image

I just pushed up a change to fix this.  But am awaiting another full round of testing.

**Comment by swirtSJW** [comment]
Yep that last change seemed to fix it.
image


@ndouglas  can you run a full round of QA on this to double check me?  (never trust a cyclops past midnight)

**Review by swirtSJW — APPROVED** [review]
I have added some code (last three commits) that resolve the incorrect setting of the changed date.  I approve of this and it tests out, but someone else needs to review my code before this is considered mergeable..

**Comment by ndouglas** [comment]
- Verified that no new log messages appeared
- looks like the changed dates aren't visibly updated, which is great!
Screen Shot 2022-12-01 at 11 33 01 AM
- checking various permutations of human vs. CMS migrator, draft vs. published vs. archived, etc, looking correct.  I was going to share screenshots from each but seemed superfluous.  One new one was this:

Screen Shot 2022-12-01 at 11 40 40 AM

- another interesting one is this one, which shows that the date is working correctly even when it's buried several revisions in the past.

Screen Shot 2022-12-01 at 11 41 50 AM

I'm not crazy about running this on Thursday night, for reasons, but I think everything is matching up.

**Review by edmund-dunn — APPROVED** [review]
This looks good to me. Solid work Nate!
